nil/services/rpc/jsonrpc/types.go

```diff
@@ -303,38 +325,44 @@ func (re *RPCReceipt) IsCommitted() bool {
 	return true
 }
 
-func NewRPCInTransaction(
-	transaction *types.Transaction, receipt *types.Receipt, index types.TransactionIndex,
-	blockHash common.Hash, blockId types.BlockNumber,
-) (*RPCInTransaction, error) {
-	hash := transaction.Hash()
-	if receipt == nil || hash != receipt.TxnHash {
-		return nil, errors.New("txn and receipt are not compatible")
-	}
-
-	result := &RPCInTransaction{
+func NewTransaction(transaction *types.Transaction) *Transaction {
+	return &Transaction{
 		Flags:                transaction.Flags,
-		Success:              receipt.Success,
 		RequestId:            transaction.RequestId,
 		Data:                 hexutil.Bytes(transaction.Data),
-		BlockHash:            blockHash,
-		BlockNumber:          blockId,
 		From:                 transaction.From,
-		GasUsed:              receipt.GasUsed,
 		FeeCredit:            transaction.FeeCredit,
 		MaxPriorityFeePerGas: transaction.MaxPriorityFeePerGas,
 		MaxFeePerGas:         transaction.MaxFeePerGas,
-		Hash:                 hash,
+		Hash:                 transaction.Hash(),
 		Seqno:                hexutil.Uint64(transaction.Seqno),
 		To:                   transaction.To,
 		RefundTo:             transaction.RefundTo,
 		BounceTo:             transaction.BounceTo,
-		Index:                hexutil.Uint64(index),
 		Value:                transaction.Value,
 		Token:                transaction.Token,
 		ChainID:              transaction.ChainId,
 		Signature:            transaction.Signature,
 	}
+}
+
+func NewRPCInTransaction(
+	transaction *types.Transaction, receipt *types.Receipt, index types.TransactionIndex,
+	blockHash common.Hash, blockId types.BlockNumber,
+) (*RPCInTransaction, error) {
+	hash := transaction.Hash()
+	if receipt == nil || hash != receipt.TxnHash {
+		return nil, errors.New("txn and receipt are not compatible")
+	}
+
+	result := &RPCInTransaction{
+		Transaction: NewTransaction(transaction),
```

Comment:
You can calculate hash only once
```suggestion
	txn := NewTransaction(transaction)
	if receipt == nil || txn.Hash != receipt.TxnHash {
		return nil, errors.New("txn and receipt are not compatible")
	}

	result := &RPCInTransaction{
		Transaction: txn,
```